File: 01/main.exs
defmodule App do
  @digits %{
    one: 1,
    two: 2,
    three: 3,
    four: 4,
    five: 5,
    six: 6,
    seven: 7,
    eight: 8,
    nine: 9
  }

  def process_file do
    {:ok, text} = File.read("./input.txt")
    lines = String.split(text, "\n")
    process_lines(lines)
  end

  def process_lines(lines, acc \\ 0)
  def process_lines([], acc), do: acc
  def process_lines([line | rest], acc) do
    digits = find_digits(line)
    new_value = acc + Enum.at(digits, 0) * 10 + Enum.at(digits, -1)
    process_lines(rest, new_value)
  end

  def find_digits(string) do
    Enum.reverse(find_digits_rev(String.graphemes(string)))
  end

  def find_digits_rev(characters, found_digits \\ [])
  def find_digits_rev([], found_digits), do: found_digits
  def find_digits_rev([head | tail] = characters, found_digits) do
    first_digit = case Integer.parse(head) do
      {digit, _} -> digit
      :error -> starts_with_digit(characters)
    end

    find_digits_rev(tail, prepend_if_not_nil(first_digit, found_digits))
  end

  def starts_with_digit(characters) do
    string = Enum.join(characters)
    Enum.find_value(@digits, fn {digit_name, digit_value} ->
      String.starts_with?(string, to_string(digit_name)) && digit_value
    end)
  end

  def prepend_if_not_nil(:nil, rest), do: rest
  def prepend_if_not_nil(head, rest), do: [head | rest]
end

IO.puts("Result: #{App.process_file()}")

File: 01/main.py
digit_map = {
    "one":      "1",
    "two":      "2", 
    "three":    "3", 
    "four":     "4", 
    "five":     "5", 
    "six":      "6", 
    "seven":    "7", 
    "eight":    "8", 
    "nine":     "9"
}

def run():
    with open('./input.txt') as input_file:
        lines = input_file.readlines()
        parsed_lines = parse_lines(lines)
        total = compute_sum(parsed_lines)
        print(f"result: {total}")

def parse_lines(input_lines):
    parsed_lines = []
    for line in input_lines:
        digits = []
        #line = line.replace('*','')
        for spelled_digit, digit in digit_map.items():
            found_index = 0
            while found_index < len(line):
                found_index = line.find(spelled_digit, found_index)
                if found_index != -1:
                    digits.append({'index': found_index, 'digit': digit})
                else:
                    break

                found_index += 1

            

        for digit in range(0,10):
            found_index = 0
            while found_index < len(line):
                found_index = line.find(str(digit), found_index)
                if found_index != -1:
                    digits.append({'index': found_index, 'digit': str(digit)})
                else:
                    break

                found_index += 1
        
        sorted_digits = sorted(digits, key=lambda a: a['index'])


        first_last = ''.join([sorted_digits[0]['digit'], sorted_digits[-1]['digit']])
        first_last_number = int(first_last)
        parsed_lines.append(first_last_number)

        print(f"{line} -> {sorted_digits} -> {first_last_number}")


    return parsed_lines

def compute_sum(char_lines):
    return sum(char_lines)

if __name__ == '__main__':
    run()